### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by segheim**

### CODE REVIEW

File: src/main/java/com/epam/esm/giftcertificatesystem/service/GiftCertificateService.java [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+import com.epam.esm.giftcertificatesystem.exception.ServiceException;
+import com.epam.esm.giftcertificatesystem.model.GiftCertificate;
+import com.epam.esm.giftcertificatesystem.model.PatchGiftCertificate;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+public interface GiftCertificateService extends GCSService<GiftCertificate> {
```
Comment (senator2202) [inline src/main/java/com/epam/esm/giftcertificatesystem/service/GiftCertificateService.java:11]: Store interfaces in package root, store there implementations in subpackage "impl".
Example: 
image

File: src/main/java/com/epam/esm/giftcertificatesystem/service/GiftCertificateServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+    private final TagRepository tagRepository;
+
+    @Autowired
+    public GiftCertificateServiceImpl(GCRepository gcsRepository, GCToTagRepository giftCertificateToTagRepository, TagRepository tagRepository) {
+        this.gcsRepository = gcsRepository;
+        this.giftCertificateToTagRepository = giftCertificateToTagRepository;
+        this.tagRepository = tagRepository;
+    }
+
+    @Transactional(rollbackFor = ServiceException.class)
```
Comment (senator2202) [inline src/main/java/com/epam/esm/giftcertificatesystem/service/GiftCertificateServiceImpl.java:35]: rollback for this exception is redundant, as there is nothing to rollback if this exception is thrown. It us just validation exception

File: src/main/java/com/epam/esm/giftcertificatesystem/model/GiftCertificate.java [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+    }
+
+    public GiftCertificate() {}
+
+    public GiftCertificate(Long id, String name, String description, BigDecimal price,
+                           Integer duration, String createDate, String lastUpdateDate) {
+        this(id, name, description, price, duration, createDate, lastUpdateDate, null);
+    }
+
+    public GiftCertificate(String name, String description, BigDecimal price,
```
Comment (senator2202) [inline src/main/java/com/epam/esm/giftcertificatesystem/model/GiftCertificate.java:44]: Remove constructors, that you don't use. The same for all entities

File: src/main/java/com/epam/esm/giftcertificatesystem/model/PatchGiftCertificate.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
+package com.epam.esm.giftcertificatesystem.model;
+
+import java.math.BigDecimal;
+import java.util.List;
+import java.util.Set;
+
+public class PatchGiftCertificate {
```
Comment (senator2202) [inline src/main/java/com/epam/esm/giftcertificatesystem/model/PatchGiftCertificate.java:7]: I don't understand why do you need this class. You can use GiftCertificate class for patch purposes

File: src/main/java/com/epam/esm/giftcertificatesystem/model/ResponseException.java [hunk 5]
```diff
@@ -0,0 +1,18 @@
+package com.epam.esm.giftcertificatesystem.model;
+
+public class ResponseException {
```
Comment (senator2202) [inline src/main/java/com/epam/esm/giftcertificatesystem/model/ResponseException.java:3]: Is it correct place for this Exception class?

File: src/main/java/com/epam/esm/giftcertificatesystem/dao/CrdRepository.java [hunk 6]
```diff
@@ -0,0 +1,12 @@
+package com.epam.esm.giftcertificatesystem.dao;
+
+import java.util.Optional;
+
+public interface CrdRepository<T> {
```
Comment (senator2202) [inline src/main/java/com/epam/esm/giftcertificatesystem/dao/CrdRepository.java:5]: Why Crd? Why not Crud? This name "CrdRepository" is not obvious

File: src/main/java/com/epam/esm/giftcertificatesystem/config/WebConfig.java [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+import javax.servlet.ServletContext;
+import javax.servlet.ServletException;
+import javax.servlet.ServletRegistration;
+
+public class WebConfig implements WebApplicationInitializer {
+
+    @Override
+    public void onStartup(ServletContext servletContext) throws ServletException {
+
+//        AnnotationConfigWebApplicationContext rootContext = new AnnotationConfigWebApplicationContext();
```
Comment (senator2202) [inline src/main/java/com/epam/esm/giftcertificatesystem/config/WebConfig.java:17]: Remove this code if you don't need it

File: src/main/java/com/epam/esm/giftcertificatesystem/config/ApplicationConfig.java [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+import javax.sql.DataSource;
+
+@Configuration
+@EnableTransactionManagement
+@ComponentScan(basePackages = "com.epam.esm.giftcertificatesystem")
+public class ApplicationConfig {
+
+    @Bean
+    public DataSource dataSource() {
+        DriverManagerDataSource dataSource = new DriverManagerDataSource();
```
Comment (senator2202) [inline src/main/java/com/epam/esm/giftcertificatesystem/config/ApplicationConfig.java:22]: Move this hardcoded datasource parameters into property file, and use properties in dataSource configuration

**Review by senator2202 — CHANGES_REQUESTED** [review]

File: src/main/java/com/epam/esm/giftcertificatesystem/web/GiftCertificateController.java [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+    private final GiftCertificateService giftCertificateService;
+
+    @Autowired
+    public GiftCertificateController(GiftCertificateService giftCertificateService) {
+        this.giftCertificateService = giftCertificateService;
+    }
+
+    @PostMapping
+    public GiftCertificate create(@RequestBody GiftCertificate giftCertificate, Errors result) throws ControllerException, ServiceException {
+        if (giftCertificate.getName() == null) {
```
Comment (senator2202) [inline src/main/java/com/epam/esm/giftcertificatesystem/web/GiftCertificateController.java:31]: Move all validation logic into separte util class and call validation logic from service layer

File: src/main/java/com/epam/esm/giftcertificatesystem/service/Impl/GiftCertificateServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+        return giftCertificates.stream()
+                .map(gc -> findById(gc.getId()).get())
+                .collect(Collectors.toList());
+    }
+
+    private boolean isValidSortType(String sortType) {
+        return sortType.equals("desc") || sortType.equals("asc");
+    }
+
+    private boolean isValidSortName(String sortName) {
```
Comment (senator2202) [inline src/main/java/com/epam/esm/giftcertificatesystem/service/Impl/GiftCertificateServiceImpl.java:77]: move validation into validation class

File: src/main/java/com/epam/esm/giftcertificatesystem/service/Impl/GiftCertificateServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+        }
+        return giftCertificate;
+    }
+
+    @Transactional
+    public Optional<GiftCertificate> delete(Long id){
+        final Optional<GiftCertificate> giftCertificate = gcsRepository.readById(id);
+        if (giftCertificate.isPresent()) {
+            giftCertificate.get().getTags()
+                .stream()
```
Comment (senator2202) [inline src/main/java/com/epam/esm/giftcertificatesystem/service/Impl/GiftCertificateServiceImpl.java:120]: .forEach() can be applied to collections without casting them tthem to Stream object.

File: src/main/java/com/epam/esm/giftcertificatesystem/model/ResponseControllerHandlerExceptionAdvice.java [hunk 12]
```diff
@@ -0,0 +1,18 @@
+package com.epam.esm.giftcertificatesystem.model;
+
+public class ResponseControllerHandlerExceptionAdvice {
```
Comment (senator2202) [inline src/main/java/com/epam/esm/giftcertificatesystem/model/ResponseControllerHandlerExceptionAdvice.java:3]: As I can see, this is simple POJO, don't name it with "Advice" suffix. It is a little bit confusing

**Review by senator2202 — CHANGES_REQUESTED** [review]

File: src/main/java/com/epam/esm/giftcertificatesystem/config/ApplicationConfig.java [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+
+    @Autowired
+    private Environment environment;
+
+    @Bean
+    public DataSource dataSource() {
+        DriverManagerDataSource dataSource = new DriverManagerDataSource();
+        dataSource.setDriverClassName(environment.getProperty("db.driver"));
+        dataSource.setUrl(environment.getProperty("db.url"));
+        dataSource.setUsername(environment.getProperty("db.name"));
```
Comment (Lavazz) [inline src/main/java/com/epam/esm/giftcertificatesystem/config/ApplicationConfig.java:31]: could you please make parameters as constants.

File: src/main/java/com/epam/esm/giftcertificatesystem/service/GCSService.java [hunk 14]
```diff
@@ -0,0 +1,15 @@
+package com.epam.esm.giftcertificatesystem.service;
+
+import com.epam.esm.giftcertificatesystem.exception.ServiceException;
+
+import java.util.Optional;
+
+public interface GCSService<T> {
+
+    Optional<T> add(T t) throws ServiceException;
```
Comment (senator2202) [inline src/main/java/com/epam/esm/giftcertificatesystem/service/GCSService.java:9]: I can not see, where Service Exception can thrown. Please double check all such cases and remove this declaration, if Exception is not thrown

**Review by senator2202 — CHANGES_REQUESTED** [review]
- Remove unused exceptions from method declarations
- Remove all commented pieces of code

**Review by senator2202 — CHANGES_REQUESTED** [review]
Please, resolve last issues

**Review by senator2202 — APPROVED** [review]

**Comment by segheim** [comment]
Thanks for your comments!

**Pull request merged**